src/ReactFinalForm.js

```diff
@@ -281,13 +271,17 @@ class ReactFinalForm extends React.Component<Props, State> {
           return this.form.reset(values)
         })
     }
-    return renderComponent(
-      {
-        ...props,
-        ...renderProps,
-        __versions: versions
-      },
-      'ReactFinalForm'
+    return React.createElement(
+      ReactFinalFormContext.Provider,
+      { value: { reactFinalForm: this.form } },
```

Comment:
this is not flat, therefore breaking any shallow comparing underneath, or is it intentional to rerender each consumer when this renders?

Reply:
Where you want to apply shallow comparing?

Reply:
There is way to overcome such issue. I should write provide with reference value not with object literal

Reply:
well, its not exactly about the shallow comparing, but the principal is the same - https://reactjs.org/docs/context.html#caveats

Reply:
Yeah, I got your idea. Then I will change PR a little for this cavet